hoodie-client/src/main/java/com/uber/hoodie/func/BufferedIterator.java

```diff
@@ -0,0 +1,224 @@
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *           http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package com.uber.hoodie.func;
+
+import com.google.common.base.Optional;
+import com.google.common.base.Preconditions;
+import com.uber.hoodie.common.model.HoodieRecord;
+import com.uber.hoodie.common.model.HoodieRecordPayload;
+import com.uber.hoodie.config.HoodieWriteConfig;
+import com.uber.hoodie.exception.HoodieException;
+import com.uber.hoodie.io.HoodieIOHandle;
+import org.apache.avro.Schema;
+import org.apache.avro.generic.IndexedRecord;
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.spark.util.SizeEstimator;
+
+import java.io.IOException;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Used for buffering input records. Buffer limit is controlled by {@link #bufferMemoryLimit}. It internally samples
+ * every {@link #RECORD_SAMPLING_RATE}th record and adjusts number of records in buffer accordingly. This is done to
+ * ensure that we don't OOM.
+ */
+public class BufferedIterator<K extends HoodieRecordPayload, T extends HoodieRecord<K>> implements Iterator<T> {
+
+  private static Logger logger = LogManager.getLogger(BufferedIterator.class);
+  private static final int RECORD_SAMPLING_RATE = 64;
+  private static final int RECORD_CACHING_LIMIT = 128 * 1024;
+
+  /**
+   * We will be using record size to determine how many records we should cache and will change permits accordingly.
+   */
+  private final Semaphore rateLimiter = new Semaphore(1);
+  private int currentRateLimit = 1;
+  private final LinkedBlockingQueue<BufferedIteratorRecord<T>> buffer = new LinkedBlockingQueue<>();
+  private final long bufferMemoryLimit;
+  private final Schema schema;
+
+  private long avgRecordSizeOfSampledRecords = 0;
+  private long numOfSampledRecords = 0;
+  private final Iterator<T> internalIterator;
+
+  private final AtomicReference<Exception> hasFailed = new AtomicReference(null);
+  private final AtomicBoolean isDone = new AtomicBoolean(false);
+  private T nextRecord;
+
+  private final AtomicLong readCounter = new AtomicLong(0);
+
+  private final AtomicLong samplingRecordCounter = new AtomicLong(-1);
+
+  public BufferedIterator(final Iterator<T> iterator, HoodieWriteConfig config) {
```

Comment:
pass in the just the value for the buffer limit instead of the whole `HoodieWriteConfig` object? A class named `BufferedIterator` having HoodieWriteConfig passed in seems like something to avoid IMO .

Reply:
along with bufferLimit I also need schema. will pass both in then.